docs/history-service.md

```diff
@@ -0,0 +1,274 @@
+# History Service
+
+This page is an introduction to the History Service architecture and internal implementation.
+See [Temporal overview](./README.md) for a system architecture overview placing this in context.
+
+<!-- https://lucid.app/lucidchart/0202e4b8-5258-4cd6-a6a0-67159300532b/edit -->
+<img src="https://github.com/temporalio/temporal/assets/52205/47a7c139-1bb4-46f9-a401-793a63e39341">
+
+The History Service handles two main types of request (gRPC) relating to an individual Workflow Execution:
+
+1. **Requests originating from the User Application** (Start / Cancel / Query / Update / Signal / Reset etc).
+
+2. **Requests originating from Temporal Workers**, on completion of a Workflow Task or Activity Task.
+
+A simplistic description of what the History Service does when handling these requests is:
+
+1. _Determine the new History Events that the request implies, and append these to Workflow History._
+
+2. _If the request implies that a Workflow or Activity Task should be created for the workflow, then add a Transfer Task in a History Service task queue, representing the task that must be added to Matching Service._
+
+3. _Alternatively, if the request is a Worker reporting that the user workflow is blocked on a timer, then add a Timer Task in a History Service task queue; when that timer expires, create a Transfer Task representing the Workflow Task in Matching that is required to continue advancing the workflow._
+
+This is illustrated in the diagram below, which contains a typical sequence of events in the execution of the following workflow pseudocode (see [Workflow Lifecycle](./workflow-lifecycle.md) for more detailed scenarios):
+
+```javascript
+function myWorkflow() {
+  result = callActivity(myActivity);
+  workflowSleep(THIRTY_DAYS);
+  return result;
+}
+```
+
+```mermaid
+sequenceDiagram
+participant app as User Application
+participant worker as Worker
+participant handler as History Shard<br>RPC Handler
+participant qp as History Shard<br>Queue Processors
+participant matching as Matching Service
+app->>handler: StartWorkflowExecution
+note right of handler: State transition:<br>initialize Workflow Execution History<br>update Mutable State<br>create Transfer Task (Workflow Task)
+qp->>matching: enqueue Workflow Task
```

Comment:
I think it would be hard to do that without damaging the clarity of this diagram. Every arrow on the right is labeled `enqueue XXX Task` and I'm hoping that that the repeated word "enqueue" will be enough to cause people to realize what is doing the dequeuing. The full cycle is depicted in the diagram at the top of the page; but this page is really documenting the History Service so I think it makes sense here to show the History Service acting as server and client, but not to show the parts of the cycle that don't involve History Service.